Add MinusLaplacian subdomain preconditioner to speed up elliptic solves

nilsvu opened the pull request:
## Proposed changes

This class approximately solves the elliptic problem on an element-centered subdomain by assuming the elliptic operator is just a Laplacian for each tensor component separately. It can be a very effective preconditioner to speed up elliptic solves, while also keeping their memory consumption low.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Support extra Schwarz subdomain preconditioners

nilsvu committed:
Support overriding boundary conditions in subdomain operator

nilsvu committed:
Add MinusLaplacian subdomain preconditioner

nilsvu committed:
Pass arguments through serial linear solvers

nilsvu committed:
Add missing includes

nilsvu commented:
@wthrowe and @moxcodes would you be willing to take this one?

wthrowe commented on src/NumericalAlgorithms/LinearSolver/ExplicitInverse.hpp:
```diff
@@ -109,10 +109,13 @@ class ExplicitInverse : public LinearSolver<LinearSolverRegistrars> {
+  template <typename LinearOperator, typename VarsType, typename SourceType,
+            typename... OperatorArgs>
+  Convergence::HasConverged solve(
+      gsl::not_null<VarsType*> solution, LinearOperator&& linear_operator,
```
Why are you taking this as a universal reference?  You don't forward it anywhere.

nilsvu replied:
This way the argument preserves `const`-ness of the value that's passed in. If the `linear_operator` has a `const` call-operator, then it may be passed in as `const`. But if the `linear_operator` has a non-`const` call-operator, then it must be passed in as non-`const`. If there's a better way to handle this, I'm happy for suggestions!

Note that the subdomain operator currently has a non-const call-operator because it keeps memory buffers as member variables. I could make those memory buffers `mutable` to make the call-operator `const`. Not sure what's preferable.

wthrowe replied:
`gsl::not_null<LinearOperator*>`

C++'s type system is weird and unintuitive, but occasionally the rules make things simpler than they look like they should be.  Here, `LinearOperator` can be inferred as `const Foo` if necessary.  (This also works with mutable lvalue references, so you can `auto& x = db::get(...)` and get confusing errors when you try to use it.)

nilsvu replied:
Does that work if the `linear_operator` is a lambda function?

wthrowe replied:
Lambdas yes, but temporaries no, which probably is an issue.  OK, leave as-is.

nikwit replied:
Memory buffer member variables are a prime example why the `mutable` keyword exists in the first place, so this with `const LinearOperator&` seems correct to me.

nilsvu replied:
I can do that and make a bunch of private memory buffers in the subdomain operator `mutable`. Yes, @wthrowe?

wthrowe replied:
This is why `mutable` members exist, and I won't block such a change, but keep in mind that it means you have to be more careful using the object in multithreaded code.  We usually assume that const methods are thread-safe, but `mutable` members change that.

Of course, (smart) pointers can also break thread safety in const methods, and those cases are harder to detect.

nilsvu replied:
Ok I guess I'll leave this as-is then, unless you ask you me make a change.

nilsvu replied:
Thinking about this again, I decided to make the `linear_operator` argument `const`. Conceptionally, applying a linear operator should not change it. The mutable memory buffers are private, they don't change publicly observable state. The subdomain operator wasn't thread-safe before, and still isn't, and to draw attention to the fact that the call-operator being const doesn't mean it's thread-safe I added a warning to the docs. Sound ok @wthrowe? Please take a look at the fixups when you get the chance.

wthrowe replied:
Yeah, that's fine.  Other fixups are fine, too.

wthrowe commented on src/NumericalAlgorithms/LinearSolver/Gmres.hpp:
```diff
@@ -355,11 +357,12 @@ Gmres<VarsType, Preconditioner, LinearSolverRegistrars>::Gmres(
 Gmres<VarsType, Preconditioner, LinearSolverRegistrars>::solve(
     const gsl::not_null<VarsType*> initial_guess_in_solution_out,
-    const LinearOperator& linear_operator, const SourceType& source,
+    LinearOperator&& linear_operator, const SourceType& source,
```
Also here.

wthrowe commented on src/NumericalAlgorithms/LinearSolver/LinearSolver.hpp:
```diff
@@ -76,7 +76,7 @@ class LinearSolver : public PUP::able {
   Convergence::HasConverged solve(
       gsl::not_null<VarsType*> initial_guess_in_solution_out,
-      const LinearOperator& linear_operator, const SourceType& source,
+      LinearOperator&& linear_operator, const SourceType& source,
```
Again.

wthrowe commented on src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp:
```diff
@@ -0,0 +1,256 @@
+                                                          OptionsGroup>;
+  using SubdomainData = ::LinearSolver::Schwarz::ElementCenteredSubdomainData<
+      Dim, tmpl::list<Poisson::Tags::Field>>;
+  using solver_type = Solver;
```
You have type aliases `SolverType` and `solver_type` that are different things.  This seems confusing.  Can you rename one of them?

wthrowe commented on src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp:
```diff
@@ -0,0 +1,256 @@
+  void reset() noexcept override { solver().reset(); }
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept override { p | solver_; }
```
Pup the base class.

wthrowe commented on src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp:
```diff
@@ -0,0 +1,256 @@
+template <size_t Dim, typename LhsTagsList, typename RhsTagsList>
+void assign_component(
+    ::LinearSolver::Schwarz::ElementCenteredSubdomainData<Dim, LhsTagsList>&
+        lhs,
```
`not_null`

wthrowe commented on src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp:
```diff
@@ -0,0 +1,256 @@
+                   subdomain_operator_, source_, operator_args);
+    detail::assign_component(*initial_guess_in_solution_out,
+                             initial_guess_in_solution_out_, component, 0);
+  };
```
No `;` on a `for`.

wthrowe reviewed (changes requested):
The current fixups look good.

nilsvu commented:
@nikwit Could you provide the second review here?

nikwit commented on src/Elliptic/DiscontinuousGalerkin/SubdomainOperator/SubdomainOperator.hpp:
```diff
@@ -232,31 +251,43 @@ struct SubdomainOperator
+                    << local_direction
+                    << " has an unexpected type. Make sure it derives off the "
+                       "'boundary_conditions_base' class set in the system.");
+            return dynamic_cast<const BoundaryConditionsBase&>(
```
not new code, but this is an up-cast, so you should be able to `static_cast`, right?

nilsvu replied:
it's a down-cast because the domain holds a generic `domain::BoundaryCondition`, and this casts it to a system-specific boundary-condition base class.

nikwit commented on src/NumericalAlgorithms/LinearSolver/ExplicitInverse.hpp:
```diff
@@ -178,7 +182,10 @@ Convergence::HasConverged ExplicitInverse<LinearSolverRegistrars>::solve(
       unit_vector_data = 1.;
       // Invoke the operator on the unit vector
-      linear_operator(make_not_null(&result_buffer), unit_vector);
+      std::apply(linear_operator,
```
should `LinearOperator` have a protocol specifying the call function signature (or at least the first two arguments of it)? If not, maybe describe in the docstrings what it needs to look like.

nilsvu replied:
A protocol doesn't work I think because the linear operator is just an invocable that's intentionally barely constrained, e.g. it can be a lambda function or a struct with a call-operator. It's documented in the `LinearSolver` base class, and I extended the docs a bit in a fixup.

nikwit commented on src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp:
```diff
@@ -0,0 +1,261 @@
+  }
+  Solver& solver() noexcept {
+    if constexpr (std::is_abstract_v<Solver>) {
+      return *solver_;
```
these public member functions have a large risk of creating a dangling reference. I would change this to return a `shared_ptr`.

nilsvu replied:
How so? They just return a reference to the object held by the unique_ptr. What am I missing?

nikwit replied:
as far as I understand, I can call this function and get the reference. Then the MinusLaplacian can get destructed which would destruct the `solver_` object because the unique pointer has no way of accounting for any references to `solver_` that may still exist. Then my reference is dangling, or not?

nilsvu replied:
Makes sense. Another basic C++ question then: How is this avoided for accessors that return a const-ref? I haven't seen accessor functions account for this sort of thing yet.

nilsvu replied:
Isn't it just a programmer error to hold refs into an object that gets destructed?

nikwit replied:
I think usually it is left to the user to make sure that all references they hold are referencing objects that still exist.

But in this case I do not like that you are taking a reference of a `unique_ptr` object, that sort of defeats the purpose (other than automatic destruction).

nilsvu replied:
https://blog.galowicz.de/2016/03/23/const_reference_to_temporary_object/
Ok I learned something today. So I just make the non-const accessor private?

nikwit replied:
I think this article is about const rvalue references which is not the case here.

wthrowe replied:
`shared_ptr` is discouraged in SpECTRE, mostly because there is no reasonable way to serialize a multiply owned object.

> I think this article is about const rvalue references which is not the case here.

Const references, I think of either type, to temporaries of various sorts.  But yes, not relevant.

> I think usually it is left to the user to make sure that all references they hold are referencing objects that still exist.

Yes.  C++ is not designed to be inherently memory safe.  As C++ programmers we have to accept that we must manually manage lifetimes.

> But in this case I do not like that you are taking a reference of a `unique_ptr` object, that sort of defeats the purpose (other than automatic destruction).

Not sure what your objection is here.  `unique_ptr<X>` is conceptually the same as a variable of type `X`, but on the heap rather than the stack.  Both have an "ownership" quality, which is the opposite of `X&`, which is access without ownership.  A reference to a `unique_ptr` object itself would be kind of odd, but that's not what is happening here.

nikwit replied:
alright, I guess that is fair, then leave it as it is. Can you delete the non-const ref accessor though once/if the buffers of the solver are mutable and the call operators are `const`?

nilsvu replied:
I made the non-const accessor private now. It's needed for the `reset` function.

nikwit commented on src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp:
```diff
@@ -0,0 +1,261 @@
+    : public LinearSolver::Serial::LinearSolver<LinearSolverRegistrars> {
+ private:
+  using Base = LinearSolver::Serial::LinearSolver<LinearSolverRegistrars>;
+  using StoredSolverType = tmpl::conditional_t<std::is_abstract_v<Solver>,
```
would it be possible to just always hold a pointer to simplify things?

nilsvu replied:
I tried that when building this class initially. The distinction is necessary for option-parsing, because `unique_ptr`s always dispatch to factory-creation, so at least the option tag and the constructor have to implement the pointer/value logic. At that point I found it easier to `std::move` the option-created object directly into the member variable and distinguish pointer/value through accessors. This also has the (very slight) advantage that pointer-dereferencing is avoided when it's not necessary.

nikwit reviewed (changes requested):
a few minor points and questions

nilsvu commented:
@nikwit Pushed two fixups in response to your comments, please take a look. Cheers!

nilsvu committed:
Make subdomain operator const

- Take the DataBox as const-ref, because no mutations
  are necessary.
- Make the call-operator const, because it mutates only
  private memory buffers, no publicly observable state.

nikwit reviewed (dismissed):
LGTM

nilsvu commented:
@wthrowe @nikwit added two more fixups that add some missing tests. Cheers!

nikwit reviewed (dismissed):
looks fine

wthrowe commented:
Squash.
